src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -97,12 +95,17 @@ protected static JsonConverter<TValue> GetValueConverter(ref WriteStack state)
                         // Read method would have thrown if otherwise.
                         Debug.Assert(reader.TokenType == JsonTokenType.PropertyName);
 
-                        state.Current.JsonPropertyNameAsString = reader.GetString();
+                        ReadOnlySpan<byte> propertyName = JsonSerializer.GetPropertyName(ref state, ref reader, options);
+                        // Copy key name to JsonPropertyName for JSON Path support in case of error.
+                        state.Current.JsonPropertyName = propertyName.ToArray();
```

Comment:
We need to remember the key name since parsing might fail deeper in the JSON and `JsonPropertyName` would be part of the JSON Path, it is not exclusive to just parse the key.

Reply:
So this means we are potentially allocating twice for every key: once for this byte array, then again to create the key instance if it is a reference type (i.e string which is the only reference type we support as a dict key). I think this is the cause of the allocations increase you noted above.

I think we should special-case string keys (and use `state.Current.JsonPropertyNameAsString` as we did before), so that we don't have this regression.